### PR TITLE
fix: KeyError: u'AWS_IAM' and add typing

### PR DESCRIPTION
**Pull request by GavinZZ**
### Description of changes
Fixing Availability Dip `KeyError: u'AWS_IAM'`.

### Description of how you validated changes
Unit and integration tests passed locally.

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [unit tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions) using:
    - [ ] Correct values
    - [ ] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Do these changes include any template validations?
    - [ ] Did the newly validated properties support intrinsics prior to adding the validations? (If unsure, please review [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) before proceeding).
        - [ ] Does the pull request ensure that intrinsics remain functional with the new validations?

### Examples?

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/api/api_generator.py [hunk 1]
```diff
@@ -1172,7 +1175,6 @@ def _set_default_authorizer(  # type: ignore[no-untyped-def]
                 default_authorizer,
                 authorizers=authorizers,
                 add_default_auth_to_preflight=add_default_auth_to_preflight,
-                api_authorizers=api_authorizers,
```
Comment (GavinZZ) [inline samtranslator/model/api/api_generator.py:1175]: No need to pass `api_authorizers`. `authorizers` already contains everything from `api_authorizers`. 
https://github.com/aws/serverless-application-model/blob/develop/samtranslator/model/api/api_generator.py#L723

File: samtranslator/model/api/api_generator.py [hunk 2]
```diff
@@ -724,12 +724,11 @@ def _add_auth(self):  # type: ignore[no-untyped-def]
 
         if authorizers:
             swagger_editor.add_authorizers_security_definitions(authorizers)  # type: ignore[no-untyped-call]
-            self._set_default_authorizer(  # type: ignore[no-untyped-call]
+            self._set_default_authorizer(
                 swagger_editor,
                 authorizers,
                 auth_properties.DefaultAuthorizer,
                 auth_properties.AddDefaultAuthorizerToCorsPreflight,
-                auth_properties.Authorizers,
```
Comment (aaythapa) [inline samtranslator/model/api/api_generator.py:732]: Why do we take out `auth_properties.Authorizers` here? Don't we need it for the method

**Review by ssenchenko — COMMENTED** [review]
Somehow we missed this case in our tests, which ld to a runtime error.
Let's add some tests, which can reproduce the situation when the key was missing and make sure it's not gonna happen again. Even unit test with mocking dependencies would be better then nothing

**Review by hoffa — APPROVED** [review]
Any tests to ensure the issue doesn't get introduced again later?

File: samtranslator/model/api/api_generator.py (hunk 2 above)
Comment (GavinZZ) [reply samtranslator/model/api/api_generator.py:732]: I have removed all use case of `auth_properties.Authorizers` inside the function. This is actually a duplicate of `authorizers` as `authorizers` build from `auth_properties.Authorizers`

**Review by aaythapa — APPROVED** [review]

**Review by aahung — APPROVED** [review]

File: tests/model/api/test_http_api_generator.py [hunk 3]
```diff
@@ -95,6 +95,48 @@ def test_auth_iam_enabled(self):
             },
         )
 
+    def test_auth_iam_enabled_with_default_iam_authorizer(self):
```
Comment (hoffa) [inline tests/model/api/test_http_api_generator.py:98]: Why not a regular transform test? These are hard to read.

Comment (ssenchenko) [reply tests/model/api/test_http_api_generator.py]: Aren't these tests more stable (not flakey)?

Comment (GavinZZ) [reply tests/model/api/test_http_api_generator.py]: Transform tests aren't flaky either. Only some of the integ tests are flaky.

**Review by ssenchenko — COMMENTED** [review]
How come these are easier to read? A lot of copy/pasting, if we change anything, need to change in 4 files?

**Comment by hoffa** [comment]
> How come these are easier to read? A lot of copy/pasting, if we change anything, need to change in 4 files?

Transforms should nearly always be deterministic, so changes to transform outputs.

**Comment by ssenchenko** [comment]
> > How come these are easier to read? A lot of copy/pasting, if we change anything, need to change in 4 files?
> 
> Transforms should nearly always be deterministic, so no changes to transform outputs.
> 
> Edit: neither is the epitome of readability to be fair. There's a few issues with unit tests for transforms, since it only tests small sections of the code. e.g. it could be possible to have a plugin that runs after the transform that breaks the transform, but the unit test would still pass.

I believe it's possible, but shouldn't it be tested in that plugin then? I'm just worried that we keep adding tests which are flakey...

**Comment by hoffa** [comment]
> > > How come these are easier to read? A lot of copy/pasting, if we change anything, need to change in 4 files?
> > 
> > 
> > Transforms should nearly always be deterministic, so no changes to transform outputs.
> > Edit: neither is the epitome of readability to be fair. There's a few issues with unit tests for transforms, since it only tests small sections of the code. e.g. it could be possible to have a plugin that runs after the transform that breaks the transform, but the unit test would still pass.
> 
> I believe it's possible, but shouldn't it be tested in that plugin then? I'm just worried that we keep adding tests which are flakey...

These are not integration tests, they're never flaky. They run the full transform code locally and compare the outputs.

**Comment by ssenchenko** [comment]
> gration tests, they're never flaky. They run the full transform code locally and

gotcha, my bad :)

**Comment by GavinZZ** [comment]
Transform tests are not flaky. It's simply comparing the output against the expected json files. 

I do agree with Chris that unit tests are highly couple to implementation details, so if a small implementation detail changes we need to spend effort fixing them.

**Review by ssenchenko — APPROVED** [review]

**Review by xazhao — APPROVED** [review]

**Pull request merged**